README.md
# let-me-do

This software makes it easy to remotely support Linux-based workstations.

Once launched with administrator rights:
 - Opens SSH port
 - Provides VNC access
 - Creates the let-me-do administrator

## Installation

### Ubuntu zesty, yakkety or xenial

```bash
sudo add-apt-repository ppa:1o1o22/let-me-do
sudo apt-get update
sudo apt-get install let-me-do
```

### Archlinux

```
yaourt -S let-me-do
```


debian/changelog
let-me-do (0.0+2-0ubuntu1) release; urgency=medium

  * Add translation

 -- Loïc Escales <L0L022@openmailbox.org>  Thu, 20 Apr 2017 10:25:47 +0200

let-me-do (0.0+1-0ubuntu1) release; urgency=medium

  * Initial release

 -- Loïc Escales <L0L022@openmailbox.org>  Thu, 13 Apr 2017 19:02:04 +0200


install.bash
#!/bin/bash

if [ "$USER" != "root" ]; then
  echo "You must be logged as root"
  exit 1
fi

case $1 in
  install)
    if [ -f "/usr/bin/apt" ]; then
      apt install openssh-server iptables xclip zenity miniupnpc x11vnc gettext cmake
    fi
    systemctl enable sshd.service
    systemctl start sshd.service
    cmake . -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX:PATH=/usr -B../let-me-do-build
    cd ../let-me-do-build || exit
    make
    make install
    update-desktop-database
    echo "done !"
    ;;
  uninstall)
    cmake . -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX:PATH=/usr -B../let-me-do-build
    cd ../let-me-do-build || exit
    make
    make uninstall
    update-desktop-database
    echo "done !"
    ;;
  *)
    echo "You must run the program with install or uninstall arg"
    ;;
esac


let_me_do.sh
#!/bin/bash

tmp_file_name="/tmp/let_me_do"
tmp_file_content="$(cat "$tmp_file_name")"

function increase {
  # for (( i = $1; i <= $2; i++ )); do
  #   echo "$i"
  #   sleep 0.04
  # done
  echo "scale=2; $1/$2*100" | bc
}

function problem {
    zenity --error --text="$1"
    echo "100"
    echo "# $1"
}

function begin_root {
  increase 3 5 &
  echo "# Redirection du port ssh"
  iptables -t nat -A PREROUTING -p tcp --dport "$port_ssh" -j REDIRECT --to-port 22
  if [ $? -ne 0 ]; then
    message="Impossible de rediriger le port ssh ($port_ssh)"
    problem "$message"
    exit
  fi

  increase 4 5 &
  echo "# Ajout de l'utilisateur let_me_do"
  useradd -r -G wheel -s /bin/bash let_me_do
  echo "let_me_do:$password" | chpasswd
  if [ $? -ne 0 ]; then
    message="Impossible d'ajouter l'utilisateur let_me_do"
    problem "$message"
    exit
  fi
}

function end_root {
  echo "# Supprime la redirection du port ssh"
  iptables -t nat -D PREROUTING -p tcp --dport "$port_ssh" -j REDIRECT --to-port 22

  echo "# Supprime l'utilisateur let_me_do"
  userdel let_me_do
}

if [ -z "$tmp_file_content" ]; then
  echo "$PPID" > "$tmp_file_name"
else
  zenity --question --title="Attention" --text="L'application est déjà ouverte" --ok-label="Ok" --cancel-label="Ouvrir quand même"
  if [ $? -eq 0 ]; then
    exit
  else
    echo "$PPID" > "$tmp_file_name"
  fi
fi

local_ip="$(ip route get 1 | awk '{print $NF;exit}')"
port_ssh="6357"
port_vnc="6358"
time_port="43200" #1 day
password="$RANDOM"
computer_name="$(hostname)"

title="Accès depuis le web"
#text_message="Votre machine est maintenant accessible depuis internet à l'adresse suivante:\n$internet_ip:$port_ssh\nLe mot de passe: $password\nElle vient d'être copiée dans le presse papier, plus qu'à envoyer à votre ange gardien !"
ok_message="Couper l'accès"
icon="emblem-web"

(
increase 0 5 &
echo "# Obtention de l'addresse ip externe"
internet_ip="$(external-ip)"
if [ -z "$internet_ip" ]; then
  message="Impossible d'obtenir l'adresse ip"
  problem "$message"
  exit
fi

increase 1 5 &
echo "# Ouverture du port ssh"
upnpc -e "ssh session of $USER on $computer_name" -a "$local_ip" "$port_ssh" "$port_ssh" tcp "$time_port" > /dev/null 2>&1
if [ $? -ne 0 ]; then
  message="Impossible d'ouvrir le port ssh ($port_ssh)"
  problem "$message"
  exit
fi

increase 2 5 &
echo "# Ouverture du port vnc"
upnpc -e "vnc session of $USER on $computer_name" -a "$local_ip" "$port_vnc" "$port_vnc" tcp "$time_port" > /dev/null 2>&1
if [ $? -ne 0 ]; then
  message="Impossible d'ouvrir le port vnc ($port_vnc)"
  problem "$message"
  exit
fi

pkexec bash -c "port_ssh=$port_ssh;password=$password;\
#$(type increase);#$(type problem);#$(type begin_root);begin_root"

echo "# Ouverture de x11vnc"
x11vnc -display "$DISPLAY" -autoport "$port_vnc" -passwd "$password" -forever -noxdamage -ssl TMP -gui tray -ncache 10 > /dev/null 2>&1 &
if [ $? -ne 0 ]; then
  message="Impossible de lancer le programme x11vnc"
  problem "$message"
  exit
fi

echo "# Copie de l'adresse dans le presse papier"
echo "$internet_ip:$port_ssh $password" | xclip -selection "clipboard"

increase 5 5 &
text_message="Votre machine est maintenant accessible depuis internet à l'adresse suivante:\n$internet_ip:$port_ssh\nLe mot de passe: $password\nL'adresse et le mot de passe viennent d'être copié dans le presse papier, plus qu'à envoyer."
echo "# $text_message"
echo "100"
) | zenity --progress --no-cancel --width=600 --title="$title" --window-icon="$icon" --icon-name="$icon" --ok-label="$ok_message"

(
echo "# Fermeture de x11vnc"
killall x11vnc

echo "# Fermeture du port ssh"
upnpc -d "$port_ssh" tcp

echo "# Fermeture du port vnc"
upnpc -d "$port_vnc" tcp

pkexec bash -c "port_ssh=$port_ssh;#$(type end_root);end_root"

echo "" > "$tmp_file_name"
echo "100"
) | zenity --progress --pulsate --auto-close --no-cancel --width=200


let_me_do.bash
#!/bin/bash
source gettext.sh

TEXTDOMAIN="let_me_do"
export TEXTDOMAIN
TEXTDOMAINDIR="@LOCALE_PATH@"
export TEXTDOMAINDIR

APP_NAME="$(gettext "Let Me Do")"

if [ "$USER" != "root" ]; then
  message="$(gettext "You must be logged as root")"
  zenity --title="$APP_NAME" --error --text="$message"
  echo "$message"
  exit 1
fi

tmp_file_name="/tmp/let_me_do"
tmp_file_content="$(cat "$tmp_file_name" 2>/dev/null)"

if [ -z "$tmp_file_content" ]; then
  echo "$PPID" > "$tmp_file_name"
else
  if zenity --title="$APP_NAME" --question \
  --text="$(gettext "Let Me Do is already open")" --ok-label="$(gettext "Ok")" \
  --cancel-label="$(gettext "Open it anyway")" 2>/dev/null; then
    exit 2
  else
    echo "$PPID" > "$tmp_file_name"
  fi
fi

function increase {
  echo "scale=2; $1/$2*100" | bc
}

function problem {
    zenity --title="$APP_NAME" --error --text="$1"
    echo "100"
    echo "# $1"
}

local_ip="$(ip route get 1 | awk '{print $NF;exit}')"
port_ssh="6357"
port_vnc="6358"
time_port="43200" #1 day
password="$RANDOM"
computer_name="$(hostname)"

if [ -z "$local_ip" ]; then
  message="$(gettext "Local IP address not found")"
  zenity --title="$APP_NAME" --error --text="$message"
  echo "$message"
  exit 3
fi

(
increase 0 5
gettext "# Obtaining the public IP address"; echo
internet_ip="$(external-ip)"
if [ -z "$internet_ip" ]; then
  problem "$(gettext "Obtaining the public IP address failed")"
  exit 4
fi

increase 1 5
eval_gettext "# Opening SSH port (\$port_ssh) on the router"; echo
if ! upnpc -e "SSH session of $USER on $computer_name" -a "$local_ip" "$port_ssh" "$port_ssh" tcp "$time_port" > /dev/null 2>&1; then
  problem "$(gettext "Opening SSH port on the router failed") ($port_ssh)"
  exit 5
fi

increase 2 5
eval_gettext "# Opening VNC port (\$port_vnc) on the router"; echo
if ! upnpc -e "VNC session of $USER on $computer_name" -a "$local_ip" "$port_vnc" "$port_vnc" tcp "$time_port" > /dev/null 2>&1; then
  problem "$(gettext "Opening VNC port on the router failed") ($port_vnc)"
  exit 6
fi

increase 3 5
eval_gettext "# Redirection of SSH port (\$port_ssh to 22)"; echo
if ! iptables -t nat -A PREROUTING -p tcp --dport "$port_ssh" -j REDIRECT --to-port 22; then
  problem "$(gettext "SSH port redirection failed") ($port_ssh)"
  exit 7
fi

increase 4 5
group="$(grep "^%.* ALL=(ALL) ALL$" /etc/sudoers | sed "s/^%\(.*\) ALL=(ALL) ALL$/\1/g")"
if [ -z "$group" ]; then
  problem "$(gettext "Administrator group not found")"
  exit 8
fi
if ! grep -q "^$group:.*" /etc/group; then
  eval_gettext "# Adding a new group: \$group"; echo
  if ! groupadd -r "$group"; then
    problem "$(gettext "The addition of a new group failed:") $group"
    exit 9
  fi
fi
gettext "# Adding the user Let Me Do"; echo
if ! useradd -r -N -G "$group" -s /bin/bash -c "$APP_NAME" let_me_do; then
  problem "$(gettext "Adding the user Let Me Do failed")"
  exit 10
fi
if ! echo "let_me_do:$password" | chpasswd; then
  problem "$(gettext "Changing Let Me Do's password failed")"
  exit 11
fi

gettext "# Starting desktop sharing (x11vnc)"; echo
x11vnc -display "$DISPLAY" -autoport "$port_vnc" -passwd "$password" -forever -noxdamage -ssl TMP -gui tray -ncache 10 > /dev/null 2>&1 &

gettext "# Copying of information in the clipboard"; echo
echo -n "ssh let_me_do@$internet_ip -p $port_ssh psswd: $password" | xclip -selection "clipboard"

increase 5 5
eval_gettext "# Your computer is now accessible from the Internet.\nIP address: \$internet_ip\nPassword: \$password\nSSH port: \$port_ssh\nVNC port: \$port_vnc\nThe information required for the connection just be copied to the clipboard."; echo
echo "100"
) | zenity --title="$APP_NAME" --width=650 --progress --no-cancel \
--ok-label="$(gettext "Close Let Me Do")" 2>/dev/null

(
gettext "# Shutting down desktop sharing (x11vnc)"; echo
killall x11vnc

eval_gettext "# Closing the SSH port (\$port_ssh) on the router"; echo
upnpc -d "$port_ssh" tcp

eval_gettext "# Closing the VNC port (\$port_vnc) on the router"; echo
upnpc -d "$port_vnc" tcp

eval_gettext "# Removing SSH port (\$port_ssh) redirection"; echo
iptables -t nat -D PREROUTING -p tcp --dport "$port_ssh" -j REDIRECT --to-port 22

gettext "# Deleting user Let Me Do"; echo
userdel let_me_do

echo "" > "$tmp_file_name"
echo "100"
) | zenity --title="$APP_NAME" --width=400 --progress --pulsate --auto-close --no-cancel 2>/dev/null


let-me-do.bash
#!/bin/bash
source gettext.sh

TEXTDOMAIN="let-me-do"
export TEXTDOMAIN
TEXTDOMAINDIR="@LOCALE_PATH@"
export TEXTDOMAINDIR

APP_NAME="$(gettext "Let Me Do")"

function set_home {
  if [ "$1" ]; then
   HOME="$(grep "^.*:.*:$1:.*$" /etc/passwd|cut -f6 -d":")"
   export HOME
  fi
}

set_home "$SUDO_UID"
set_home "$PKEXEC_UID"

if [ "$USER" != "root" ]; then
  message="$(gettext "You must be logged as root")"
  zenity --title="$APP_NAME" --error --text="$message"
  echo "$message"
  exit 1
fi

tmp_file_name="/tmp/let-me-do"
tmp_file_content="$(cat "$tmp_file_name" 2>/dev/null)"

if [ -z "$tmp_file_content" ]; then
  echo "$PPID" > "$tmp_file_name"
else
  if zenity --title="$APP_NAME" --question \
  --text="$(gettext "Let Me Do is already open")" --ok-label="$(gettext "Ok")" \
  --cancel-label="$(gettext "Open it anyway")" 2>/dev/null; then
    exit 2
  else
    echo "$PPID" > "$tmp_file_name"
  fi
fi

function increase {
  echo "scale=2; $1/$2*100" | bc
}

function problem {
    zenity --title="$APP_NAME" --error --text="$1"
    echo "100"
    echo "# $1"
}

local_ip="$(ip route get 1 | awk '{print $NF;exit}')"
port_ssh="6357"
port_vnc="6358"
time_port="43200" #1 day
password="$RANDOM"
computer_name="$(hostname)"

if [ -z "$local_ip" ]; then
  message="$(gettext "Local IP address not found")"
  zenity --title="$APP_NAME" --error --text="$message"
  echo "$message"
  exit 3
fi

(
increase 0 5
gettext "# Obtaining the public IP address"; echo
internet_ip="$(external-ip)"
if [ -z "$internet_ip" ]; then
  problem "$(gettext "Obtaining the public IP address failed")"
  exit 4
fi

increase 1 5
eval_gettext "# Opening SSH port (\$port_ssh) on the router"; echo
if ! upnpc -e "SSH session of $USER on $computer_name" -a "$local_ip" "$port_ssh" "$port_ssh" tcp "$time_port" > /dev/null 2>&1; then
  problem "$(gettext "Opening SSH port on the router failed") ($port_ssh)"
  exit 5
fi

increase 2 5
eval_gettext "# Opening VNC port (\$port_vnc) on the router"; echo
if ! upnpc -e "VNC session of $USER on $computer_name" -a "$local_ip" "$port_vnc" "$port_vnc" tcp "$time_port" > /dev/null 2>&1; then
  problem "$(gettext "Opening VNC port on the router failed") ($port_vnc)"
  exit 6
fi

increase 3 5
eval_gettext "# Redirection of SSH port (\$port_ssh to 22)"; echo
if ! iptables -t nat -A PREROUTING -p tcp --dport "$port_ssh" -j REDIRECT --to-port 22; then
  problem "$(gettext "SSH port redirection failed") ($port_ssh)"
  exit 7
fi

increase 4 5
group="$(grep "^%.* ALL=(ALL) ALL$" /etc/sudoers | sed "s/^%\(.*\) ALL=(ALL) ALL$/\1/g")"
if [ -z "$group" ]; then
  problem "$(gettext "Administrator group not found")"
  exit 8
fi
if ! grep -q "^$group:.*" /etc/group; then
  eval_gettext "# Adding a new group: \$group"; echo
  if ! groupadd -r "$group"; then
    problem "$(gettext "The addition of a new group failed:") $group"
    exit 9
  fi
fi
gettext "# Adding the user Let Me Do"; echo
if ! useradd -r -N -G "$group" -s /bin/bash -c "$APP_NAME" let-me-do; then
  problem "$(gettext "Adding the user Let Me Do failed")"
  exit 10
fi
if ! echo "let-me-do:$password" | chpasswd; then
  problem "$(gettext "Changing Let Me Do's password failed")"
  exit 11
fi

gettext "# Starting desktop sharing (x11vnc)"; echo
x11vnc -display "$DISPLAY" -autoport "$port_vnc" -passwd "$password" -forever -noxdamage -ssl TMP -gui tray -ncache 10 > /dev/null 2>&1 &

information="ssh let_me_do@$internet_ip -p $port_ssh psswd: $password"

gettext "# Copying of information in the clipboard"; echo
echo -n "$information" | xclip -selection "clipboard"

if [ -e "$HOME/.let_me_do_masters" ]; then
  masters="$(cat "$HOME/.let_me_do_masters")"
  if [ "$masters" ]; then
    gettext "# Wait for reply to if you want to send an email to your master (s)"; echo
    if zenity --title="$APP_NAME" --question --icon-name="emblem-mail" --text="$(gettext "Would you like to send your connection information by email to your master(s) ?")"; then
      xdg-open "mailto:$masters?subject=Let Me Do at $HOSTNAME&body=$information" &
    fi
  fi
fi

increase 5 5
eval_gettext "# Your computer is now accessible from the Internet.\nIP address: \$internet_ip\nPassword: \$password\nSSH port: \$port_ssh\nVNC port: \$port_vnc\nThe information required for the connection just be copied to the clipboard."; echo
echo "100"
) | zenity --title="$APP_NAME" --width=650 --progress --no-cancel \
--ok-label="$(gettext "Close Let Me Do")" 2>/dev/null

(
gettext "# Shutting down desktop sharing (x11vnc)"; echo
killall x11vnc

eval_gettext "# Closing the SSH port (\$port_ssh) on the router"; echo
upnpc -d "$port_ssh" tcp

eval_gettext "# Closing the VNC port (\$port_vnc) on the router"; echo
upnpc -d "$port_vnc" tcp

eval_gettext "# Removing SSH port (\$port_ssh) redirection"; echo
iptables -t nat -D PREROUTING -p tcp --dport "$port_ssh" -j REDIRECT --to-port 22

gettext "# Deleting user Let Me Do"; echo
userdel let-me-do

echo "" > "$tmp_file_name"
echo "100"
) | zenity --title="$APP_NAME" --width=400 --progress --pulsate --auto-close --no-cancel 2>/dev/null
